iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/pipe/task/subtask/PipeConfigSubtask.java

```diff
@@ -0,0 +1,270 @@
+    Map<String, String> allAttributes = new HashMap<>();
+    allAttributes.putAll(extractorAttributes);
+    allAttributes.putAll(connectorAttributes);
+    this.outputPipeConnector.validate(
+        new PipeParameterValidator(new PipeParameters(allAttributes)));
+  }
+
+  public void bindExecutors(
+      ListeningExecutorService subtaskWorkerThreadPoolExecutor,
+      ExecutorService subtaskCallbackListeningExecutor,
+      PipeConfigSubtaskScheduler subtaskScheduler) {
+    this.subtaskWorkerThreadPoolExecutor = subtaskWorkerThreadPoolExecutor;
+    this.subtaskCallbackListeningExecutor = subtaskCallbackListeningExecutor;
+    this.subtaskScheduler = subtaskScheduler;
+  }
+
+  @Override
+  public Boolean call() throws Exception {
+    boolean hasAtLeastOneEventProcessed = false;
+
+    try {
+      // If the scheduler allows to schedule, then try to consume an event
+      while (subtaskScheduler.schedule()) {
+        // If the event is consumed successfully, then continue to consume the next event
+        // otherwise, stop consuming
+        if (!executeOnce()) {
+          break;
+        }
+        hasAtLeastOneEventProcessed = true;
+      }
+    } finally {
+      // Reset the scheduler to make sure that the scheduler can schedule again
+      subtaskScheduler.reset();
+    }
+
+    return hasAtLeastOneEventProcessed;
+  }
+
+  /** Should be synchronized with {@link PipeConfigSubtask#releaseLastEvent} */
+  private synchronized void setLastEvent(Event event) {
+    lastEvent = event;
+  }
+
+  /**
+   * Try to consume an event by the pipe plugin.
+   *
+   * @return true if the event is consumed successfully, false if no more event can be consumed
+   * @throws Exception if any error occurs when consuming the event
+   */
+  @SuppressWarnings("squid:S112") // Allow to throw Exception
+  private boolean executeOnce() throws Exception {
+    if (isClosed.get()) {
+      return false;
+    }
+
+    final Event event = lastEvent != null ? lastEvent : null; // TODO: get an event from upstream
+    // Record the last event for retry when exception occurs
+    setLastEvent(event);
+    // TODO: process the event if necessary
+
+    try {
+      if (event == null) {
+        return false;
+      }
+
+      if (event instanceof TabletInsertionEvent) {
+        outputPipeConnector.transfer((TabletInsertionEvent) event);
+      } else if (event instanceof TsFileInsertionEvent) {
+        outputPipeConnector.transfer((TsFileInsertionEvent) event);
+      } else {
+        outputPipeConnector.transfer(event);
+      }
+
+      releaseLastEvent(true);
+    } catch (PipeConnectionException e) {
+      if (!isClosed.get()) {
+        throw e;
+      } else {
+        LOGGER.info("PipeConnectionException in pipe transfer, ignored because pipe is dropped.");
+        releaseLastEvent(false);
+      }
+    } catch (Exception e) {
+      if (!isClosed.get()) {
+        throw new PipeException(
```

Comment:
The exception message too needs to be updated since the users won't implement the connector..